README.md
**NOTE:** All 0.x releases are kind of "alpha-versions" without expectations of interface backward-compatibility.

## Table of Contents
[mdz_ansi Overview](#mdz_ansi-Overview)<br>
[mdz_ansi Advantages](#mdz_ansi-Advantages)<br>
[Performance Comparison](#performance-comparison)<br>
[mdz_ansi Usage](#mdz_ansi-Usage)<br>
[Licensing info](#Licensing-info)<br>
[Credits](#Credits)

## mdz_ansi Overview
Wiki: [mdz_ansi Wiki]<br>
file: *"mdz_ansi.h"*

Please take a look at *"mdz_ansi.h"* file or [mdz_ansi Wiki] site for detailed functions descriptions.

[mdz_ascii Wiki]: https://github.com/maxdz-gmbh/mdz_ansi/wiki/mdz_ansi-overview

[mdz_ansi] - is a very lightweight, versatile and speedy C  library for handling 1-byte (ASCII/ANSI) strings, developed by [maxdz Software GmbH]. Source code of library is highly-portable, conforms to ANSI C 89/90 Standard.

Only shared/dynamically loaded libraries (*.so* and *.dll* files with import libraries) are available for evaluation testing purposes. Static libraries are covered by our commercial licenses.

**Linux** binaries are built against Linux Kernel 2.6.18 - and thus should be compatible with Debian (from ver. 4.0), Ubuntu (from ver. 8.04), Fedora (from ver. 9)

**FreeBSD** binaries - may be used from FreeBSD ver. 7.0

**Win32** binaries are built using Visual Studio Platform Toolset "v90", thus are compatible with Windows versions from Windows 2000.<br>
**Win64** binaries are built using Visual Studio Platform Toolset "v100", thus are compatible with Windows versions from Windows XP.<br>

**Android** x86/armeabi-v7a binaries - may be used from Android API level 16 ("Jelly Bean" ver. 4.1.x)<br>
**Android** x86_64/arm64-v8a binaries - may be used from Android API level 21 ("Lollipop" ver. 5.0)

**macOS** binaries - x86_64, from *MacOS X v10.6.0*

[mdz_ansi]: https://github.com/maxdz-gmbh/mdz_ansi
[maxdz Software GmbH]: https://maxdz.com/

## mdz_ansi Advantages

**1. High portability:** the whole code conforms to ANSI C 89/90 Standard. Multithreading/asynchronous part is POSIX compatible (under UNIX/Linux).

**2. Little dependencies:** basically *mdz_containers* functions are only dependend on standard C-library memory-management/access functions. Multithreading part is dependend on POSIX *pthreads* API (under UNIX/Linux) and old process control/synchronization API (from Windows 2000). It means you can use library in your code withouth any further dependencies except standard plattform libraries/APIs.

**3. Fast:** comparison tables for *mdz_ansi_find()*, *mdz_ansi_firstOf()* are here [Performance Comparison](#performance-comparison). There will be more tables/info later.

**4. Flexibilty:** nearly all functions contain "left position" and "right position" parameters, to limit processed area from left and right. "ANSI" string contains more functions than according *STL*, *boost* or *glib* analogs have.

**5. Extended error-checking:** all functions preserve internal error-code pointing the problem. It is possible to use strict error-checking (when all preserved error-codes should be *MDZ_ERROR_NONE*) or "relaxed"-checking - when only returned *mdz_false* will indicate error.

**6. Extended control:** containers do only explicit operations. It means for example, when "insert" function is called with auto-reservation flag set in *mdz_false* - it will return error if there is not enough capacity in container. No implicit reservations will be made.

**7. Attached usage:** containers should not necessarily use dynamically-allocated memory - which may be not available on your embedded system (or if malloc()/free() are forbidden to use in you safety-critical software). Just attach container/data to your statically-allocated memory and use all containers functionality.

**8. Cache-friendly:** it is possible to keep controlling and data parts together in memory using "embedded part".

**9. Asynchronous execution:** almost all functions can be executed asynchronously

## Performance Comparison

Performance comparison tables for *[mdz_ansi_find]*() and *[mdz_ansi_firstOf]*() give an idea about *mdz_ansi* library overall performance on different platforms compared to STL and standard C library. Modern implementationsof STL and standard C library are pretty fast, using optimized versions of memory-access functions.

- **[mdz_ansi_find]() Test**

Following tests are executed:

- Test *1/100M*": Find 1 byte - in the end of 100M bytes long string<br>
- Test *5/100M*": Find 5 bytes long string - in the end of 100M bytes long string<br>
- Test *10/100M*": Find 10 bytes long string - in the end of 100M bytes long string<br>
- Test *100/100M*": Find 100 bytes long string - in the end of 100M bytes long string<br>
- Test *1K/100M*": Find 1K bytes long string - in the end of 100M bytes long string<br>
- Test *500K/1M*": Find 500K bytes long string - in the end of 1M bytes long string<br>
- Test *100M-100/100M*": Find "100M minus 100" bytes long string - in the end of 100M bytes long string<br>
- Test *100M/100M*": Find 100M bytes long string - in 100M bytes long string<br>

For **Windows 10 (64-bit)** on *Intel i5-6600 @ 3.30GHz (4 cores/4 threads)*<br>

**Monotone test** : "long string" and "string to find" are both filled with '1's; on the last position of both strings is '2'

monotone = MDZ_FIND_MONOTONE method<br>
clib = MDZ_FIND_CLIB method (standard C library)<br>
brute = MDZ_FIND_BRUTE method<br>
bmh = MDZ_FIND_BMH method

- VC++ toolset v140 (32-bit)<br>
(all numbers are in microseconds measured using *QueryPerformanceCounter()* in main execution thread)

| Test  | mdz_ansi, monotone | mdz_ansi, clib | mdz_ansi, brute |  mdz_ansi, bmh |std::string.find() | clib (strstr())|
| :---:| ---: | ---: | ---: | ---: | ---: | ---: |
| 1/100M| 70,351 |  |  |  | 162,681 | 70,579 |
| 5/100M| 407,668 | 460,052 | 3,045,869 | 781,655 | 3,381,061 | 482,075 |
| 10/100M| 1,334,782 | 707,712 | 4,394,022 | 780,128 | 4,206,329 | 731,395 |
| 100/100M| 1,333,516 | 10,914,646 | 15,779,350 | 781,370 | 15,652,407 | 11,253,026 |
| 1K/100M| 1,332,838 | 70,179,989 | 139,398,637 | 781,439 | 139,808,212 | 75,808,535 |
| 500K/1M| 13,202 | 166,409,422 | 323,375,345 | 9,411 | 324,276,637 | 178,302,908 |
| 100M-100/100M| 1,262,919 | 10,884,012 | 14,182,350 | 1,066,737 | 14,150,110 | 10,383,086 |
| 100M/100M| 117,970 | |  |  | 144,573 | 114,565 |

- MinGW/gcc toolset (32-bit)<br>
(all numbers are in microseconds measured using *QueryPerformanceCounter()* in main execution thread)

| Test  | mdz_ansi, monotone | mdz_ansi, clib | mdz_ansi, brute |  mdz_ansi, bmh |
| :---:| ---: | ---: | ---: | ---: |
| 1/100M| 148,067 |  |  |  |
| 5/100M| 534,070 | 1,599,882 | 6,825,862 | 784,326 |
| 10/100M| 551,404 | 3,635,378 | 7,898,385 | 783,832 |
| 100/100M| 550,701 | 32,447,796 | 20,451,496 | 786,006 |
| 1K/100M| 551,213 | 348,052,489 | 117,762,194 | 784,335 |
| 500K/1M| 7,851 | 814,620,053 | 246,574,213 | 6,263 |
| 100M-100/100M| 997,729 | 33,028,357 | 11,705,985| 456,680 |
| 100M/100M| 328,564 | |  |  |

- **[mdz_ansi_firstOf]() Test**

Following tests are executed:

- Test *1/100M*": Find first of 1 byte - in the end of 100M bytes long string<br>
- Test *5/100M*": Find first of 5 bytes - in the end of 100M bytes long string<br>
- Test *20/100M*": Find first of 20 bytes - in the end of 100M bytes long string<br>
- Test *50/100M*": Find first of 50 bytes - in the end of 100M bytes long string<br>
- Test *100/100M*": Find first of 100 bytes - in the end of 100M bytes long string<br>

For **Windows 10 (64-bit)** on *Intel i5-6600 @ 3.30GHz (4 cores/4 threads)*<br>

- VC++ toolset v140 (32-bit)<br>
(all numbers are in microseconds measured using *QueryPerformanceCounter()* in main execution thread)

| Test  | mdz_ansi| std::string.find_first_of() | clib (strcspn())|
| :---:| ---: | ---: | ---: |
| 1/100M| 70,078 | 163,666 | 2,085,714 |
| 5/100M| 370,204 | 3,719,660 | 2,077,677 |
| 20/100M| 369,162 | 5,714,212 | 2,076,031 |
| 50/100M| 368,994 | 10,965,401 | 2,078,038 |
| 100/100M| 369,360 | 18,727,283 | 2,076,740 |

- MinGW/gcc toolset (32-bit)<br>
(all numbers are in microseconds measured using *QueryPerformanceCounter()* in main execution thread)

| Test  | mdz_ansi|
| :---:| ---: |
| 1/100M| 153,511 |
| 5/100M| 278,387 |
| 20/100M| 276,389 |
| 50/100M| 275,956 |
| 100/100M| 277,709 |

[mdz_ansi_find]: https://github.com/maxdz-gmbh/mdz_ansi/wiki/mdz_ansi_find_async
[mdz_ansi_firstOf]: https://github.com/maxdz-gmbh/mdz_ansi/wiki/mdz_ansi_firstOf_async

## mdz_ansi Usage

**ansi** is implemented with strict input parameters checking. It means *mdz_false* or some other error indication will be returned if one or several input parameters are invalid - even if such an invalidity doesn't lead to inconsistence (for example adding or removing 0 items).<br>

**Test license generation:** - in order to get free test-license, please proceed to our Shop page [maxdz Shop] and register an account. After registration you will be able to obtain free 14-days test-licenses for our products using "Obtain for free" button. 
Test license data should be used in *mdz_ansi_init()* call for library initialization.

**NOTE:** All 0.x releases are kind of "beta-versions" and can be used 1) only with test-license (during test period of 14 days, with necessity to re-generate license for the next 14 days test period) and 2) without expectations of interface backward-compatibility.

Several usage-scenarios are possible:
- low-level - raw C interface, using *mdz_ansi.h* header file
- higher-level - using *MdzAnsi* C++ "wrapper" around *mdz_ansi.h* functions

[mdz_ansi Wiki]: https://github.com/maxdz-gmbh/mdz_ansi/wiki/mdz_ansi-overview
[Glib]: https://en.wikipedia.org/wiki/GLib
[STL]: https://en.wikipedia.org/wiki/Standard_Template_Library
[Performance Comparison]: #performance-comparison
[maxdz Shop]: https://maxdz.com/shop.php

#### Code Example (low-level use)

*mdz_ansi_init()* with license information should be called for library initialization before any subsequent calls:

```
#include <mdz_ansi.h>

int main(int argc, char* argv[])
{
  /* mdz_ansi library initialization using test info retrieved after license generation (see "Test license generation" above) */
  
  mdz_bool bRet = mdz_ansi_init("<first-name-hash>", "<last-name-hash>", "<email-hash>", "<license-hash>");
  ...
  
  mdz_ansi_uninit(); /* call for un-initialization of library */
  
  return 0;
}
```

[mdz_ansi_create]: https://github.com/maxdz-gmbh/mdz_ansi/wiki/mdz_ansi_create
[mdz_ansi_destroy]: https://github.com/maxdz-gmbh/mdz_ansi/wiki/mdz_ansi_destroy

After library initialization call *[mdz_ansi_create]*() for **ansi** creation. There should be also symmetric *[mdz_ansi_destroy]*() call for every create, otherwise allocated for **ansi** memory remains occupied:

```
#include <mdz_ansi.h>

int main(int argc, char* argv[])
{
  mdz_bool bRet = mdz_ansi_init("<first-name-hash>", "<last-name-hash>", "<email-hash>", "<license-hash>");   /* initialize pAnsi */
  
  mdz_Ansi* pAnsi = mdz_ansi_create(0); /* create ansi-string */
  ...
  ...
  /* use pAnsi */
  ...
  ...
  /* destroy pAnsi */
  
  mdz_ansi_destroy(&pAnsi); /* after this pAnsi should be NULL */
  
  mdz_ansi_uninit(); /* uninitialize pAnsi */

  ...
}
```

Use *mdz_Ansi** pointer for subsequent library calls:

```
#include <mdz_ansi.h>

int main(int argc, char* argv[])
{
  mdz_bool bRet = mdz_ansi_init("<first-name-hash>", "<last-name-hash>", "<email-hash>", "<license-hash>");
  
  mdz_Ansi* pAnsi = mdz_ansi_create(0); /* create ansi-string */

  /* reserve 5 elements, set them into 'a'. After this Capacity of string is 6 (includes terminating 0) and Size is 5. */
  
  bRet = mdz_ansi_reserveAndInit(pAnsi, 5, 'a'); /* "aaaaa" after this call */
  
  /* insert 'b' in front position with auto-reservation if necessary */
  
  bRet = mdz_ansi_insert(pAnsi, 0, "b", 1, mdz_true); /* "baaaaa" after this call */
  
  /* append string with "cde" with auto-reservation if necessary */
  
  bRet = mdz_ansi_insert(pAnsi, pAnsi->m_nSize - 1, "cde", 3, mdz_true); /* "baaaaacde" after this call */
  
  ...
  
  mdz_ansi_destroy(&pAnsi);
  
  mdz_ansi_uninit();
  ...
}
```

#### Code Example (higher-level use)

This is an example above using *MdzAnsi* C++ "wrapper":

```
#include <MdzAnsi.h>

int main(int argc, char* argv[])
{
  mdz_ansi_init("<first-name-hash>", "<last-name-hash>", "<email-hash>", "<license-hash>");
  
  MdzAnsi oAnsi; // initialize ansi-string

  bool bRet = oAnsi.reserveAndInit(5, 'a'); // "aaaaa" after this call
  
  bRet = oAnsi.insertAndReserve(0, 'b'); // "baaaaa" after this call
  
  bRet = oAnsi.appendAndReserve("cde"); // "baaaaacde" after this call
  
  mdz_ansi_uninit();
  ...
}
```

## Licensing info

Use of **mdz_ansi** library is regulated by license agreement in *LICENSE.txt*

Basically private non-commercial "test" usage is unrestricted. Commercial usage of library (incl. its source code) will be regulated by according license agreement.


HISTORY.txt
05.03.2021 (fri): Release 0.3
-----------------------------
- minor changes (comments)

01.03.2021 (mon): Release 0.2
-----------------------------
- correction in "attach" functions

28.01.2021 (thu): Release 0.1
-----------------------------

Added functions:

- mdz_ansi_init
- mdz_ansi_init_attached
- mdz_ansi_uninit

- mdz_ansi_create
- mdz_ansi_create_attached
- mdz_ansi_destroy

- mdz_ansi_clear
- mdz_ansi_attachData
- mdz_ansi_reserve
- mdz_ansi_reserveAndInit_async

- mdz_ansi_capacity
- mdz_ansi_size
- mdz_ansi_resize
- mdz_ansi_offsetFromStart
- mdz_ansi_isAttachedData
- mdz_ansi_embedSize

- mdz_ansi_insert_async

- mdz_ansi_findSingle_async
- mdz_ansi_find_async
- mdz_ansi_firstOf_async
- mdz_ansi_firstNotOf_async
- mdz_ansi_rfindSingle_async
- mdz_ansi_rfind_async
- mdz_ansi_lastOf_async
- mdz_ansi_lastNotOf_async

- mdz_ansi_removeFrom_async
- mdz_ansi_remove_async

- mdz_ansi_trimLeft_async
- mdz_ansi_trimRight_async
- mdz_ansi_trim_async

- mdz_ansi_compare_async

- mdz_ansi_replace_async

- mdz_ansi_count_async

- mdz_ansi_copySubAnsi_async
- mdz_ansi_copySubAnsiFrom_async



mdz_types.h
/**
 * \ingroup mdz libraries
 *
 * \author maxdz Software GmbH
 *
 * \par license
 * This file is subject to the terms and conditions defined in file 'LICENSE.txt', which is part of this source code package.
 *
 * \par description
 * Public types of mdz_containers library
 *
 * \par info
 * See additional info on mdz_containers library like version, portability, etc in mdz_containers.h
 */

#ifndef MDZ_TYPES_H
#define MDZ_TYPES_H

#include <stddef.h>

#include <stdint.h>
#include <limits.h>

#ifndef _WIN32
#include <pthread.h>
#endif

 /**
 * Boolean type of mdz_containers library.
 */
#define mdz_false 0
#define mdz_true 1
typedef unsigned char mdz_bool;

/**
 * 3-state boolean
 */
enum mdz_3state
{
  /**
  * False
  */
  MDZ_3STATE_FALSE = 0,

  /**
   * True
   */
  MDZ_3STATE_TRUE = 1,

  /**
  * Error
  */
  MDZ_3STATE_ERROR = -1
};

#ifdef _WIN32
typedef void* HANDLE;
#endif

/**
* Shared data for asynchronous calls
*/
struct mdz_asyncData
{
  /**
  * Pointer to string instance
  */
  void* m_pString;

  /**
  * Result of call. Invalid if call is not completely finished (m_bFinished is mdz_false)
  */
  size_t m_nResult;

  /**
  * Additional data returned by call (if any). Invalid if call is not completely finished (m_bFinished is mdz_false)
  */
  void* m_pData;

  /**
  * Thread on which the call is executed. May be used by client for wait operations
  */
#ifdef _WIN32
  HANDLE m_hThread;
#else
  pthread_t m_hThread;
#endif

  /**
  * mdz_true if the call is completely finished. Otherwise mdz_false (if interrupted/cancelled)
  */
  mdz_bool m_bFinished;

  /**
  * Should be set in mdz_true during call execution, to cancel the call. Otherwise mdz_false
  */
  mdz_bool m_bCancel;
};

/**
 * Attachment type
 */
enum mdz_attach_type
{
  /**
   * Size is 0 after attachment
   */
  MDZ_ATTACH_ZEROSIZE = 0,

  /**
   * Size is Capacity minus 1 after attachment; 0-terminator is present
   */
  MDZ_ATTACH_SIZE_TERMINATOR = 1,

  /**
   * Size is Capacity after attachment; without 0-terminator
   */
  MDZ_ATTACH_SIZE_NO_TERMINATOR = 2
};

/**
  * Data endianness
  */
enum mdz_endianness
{
  /**
   * Undefined endianness
   */
  MDZ_ENDIAN_UNDEFINED = 0,

  /**
   * Little-endian
   */
  MDZ_ENDIAN_LITTLE = 1,

  /**
   * Big-endian
   */
  MDZ_ENDIAN_BIG = 2,

  /**
   * Error endianness
   */
  MDZ_ENDIAN_ERROR = 3
};

/**
 * Result of comparison
 */
enum mdz_compare_result
{
  /**
   * Equal
   */
  MDZ_COMPARE_EQUAL = 0,

  /**
   * Non-equal
   */
  MDZ_COMPARE_NONEQUAL = 1,

  /**
   * Greater
   */
  MDZ_COMPARE_GREATER = 2,

  /**
   * Smaller
   */
  MDZ_COMPARE_SMALLER = 3,

  /**
   * Comparison ended with error
   */
  MDZ_COMPARE_ERROR = 4
};

/**
 * Method of find
 */
enum mdz_find_method
{
  /*
   * standard C-library implementation
   */
  MDZ_FIND_CLIB = 0,
  
  /**
   * Optimized Boyer-Moore-Horspool search
   */
  MDZ_FIND_BMH = 1,

  /**
   * Improved "brute-force" search
   */
  MDZ_FIND_BRUTE = 2,

  /**
   * Optimized search for monotone sub-string in monotone string
   */
  MDZ_FIND_MONOTONE = 3
};

enum mdz_error_code
{
  /**
   * No error
   */
  MDZ_ERROR_NONE = 0,

  /**
   * Invalid "data" parameter
   */
  MDZ_ERROR_DATA = 1,

  /**
   * Not enough capacity
   */
  MDZ_ERROR_CAPACITY = 2,

  /**
   * Invalid "offset" parameter
   */
  MDZ_ERROR_OFFSET = 3,

  /**
   * "Count" parameter is zero
   */
  MDZ_ERROR_ZEROCOUNT = 4,

  /**
   * "Count" parameter is too big
   */
  MDZ_ERROR_BIGCOUNT = 5,

  /**
   * "Left" parameter is too big
   */
  MDZ_ERROR_BIGLEFT = 6,

  /**
   * "Right" parameter is too big
   */
  MDZ_ERROR_BIGRIGHT = 7,

  /**
   * Invalid "items" parameter
   */
  MDZ_ERROR_ITEMS = 8,

  /**
   * Container is empty
   */
  MDZ_ERROR_EMPTY = 9,

  /**
   * Container is not empty
   */
  MDZ_ERROR_NONEMPTY = 10,

  /**
   * Invalid "sub-container" parameter
   */
  MDZ_ERROR_SUBCONTAINER = 11,

  /**
   * Container is attached
   */
  MDZ_ERROR_ATTACHED = 12,

  /**
   * Memory allocation error
   */
  MDZ_ERROR_ALLOCATION = 13,

  /**
   * Data content error
   */
  MDZ_ERROR_CONTENT = 14,

  /**
   * Invalid endiannes type
   */
  MDZ_ERROR_ENDIANNESS = 15,

  /**
   * Invalid attach typpe
   */
  MDZ_ERROR_ATTACHTYPE = 16,

  /**
   * Invalid find method
   */
  MDZ_ERROR_FINDMETHOD = 17,

  /**
   * Memory allocation for thread error
   */
  MDZ_ERROR_THREAD_ALLOC = 18,

  /**
   * Thread start error
   */
  MDZ_ERROR_THREAD_START = 19,

  /**
  * Invalid "source string" parameter
  */
  MDZ_ERROR_SOURCE = 20
};

/**
* Handle of ANSI string in mdz_string library
*/
struct mdz_Ansi
{
  /**
  * Pointer to data of string. Do not change it directly.
  */
  char* m_pData;

  /**
  * Error code if operation failed or there are some inconsistences.
  */
  enum mdz_error_code m_enErrorCode;
};

/**
* Handle of utf8 string in mdz_string library
*/
struct mdz_Utf8
{
  /**
  * Pointer to data of string. Do not change it directly.
  */
  unsigned char* m_pData;

  /**
  * Error code if operation failed or there are some inconsistences.
  */
  enum mdz_error_code m_enErrorCode;
};

/**
* Handle of UTF-16 string in mdz_string library
*/
struct mdz_Utf16
{
  /**
  * Pointer to data of string. Do not change it directly.
  */
  const uint16_t* m_pData;

  /**
  * Error code if operation failed or there are some inconsistences.
  */
  enum mdz_error_code m_enErrorCode;
};

/**
* Handle of UTF-32 string in mdz_string library
*/
struct mdz_Utf32
{
  /**
  * Pointer to data of string. Do not change it directly.
  */
  const uint32_t* m_pData;

  /**
  * Error code if operation failed or there are some inconsistences.
  */
  enum mdz_error_code m_enErrorCode;
};

/**
* Handle of "wide"-character string in mdz_string library
*/
struct mdz_Wchar
{
  /**
  * Pointer to data of string. Do not change it directly.
  */
  const wchar_t* m_pData;

  /**
  * Error code if operation failed or there are some inconsistences.
  */
  enum mdz_error_code m_enErrorCode;
};

#endif
